src/BreakpointViewer.cpp

```diff
@@ -0,0 +1,621 @@
+#include <QPainter>
+#include <QPaintEvent>
+#include <QMessageBox>
+#include <QToolTip>
+#include <QComboBox>
+#include <QDebug>
+#include <cassert>
+#include "BreakpointViewer.h"
+#include "ui_BreakpointViewer.h"
+#include "Convert.h"
+#include "CommClient.h"
+#include "OpenMSXConnection.h"
+
+enum TableColumns {
+	ENABLED = 0,
+	WP_TYPE = 1,
+	LOCATION = 2,
+	BP_ADDRESS = 2,
+	WP_REGION = 2,
+	CONDITION = 3,
+	SLOT = 4,
+	SEGMENT = 5,
+};
+
+struct ItemPos {
+	BreakpointViewer::Type type;
+	int row;
+	bool operator==(const ItemPos &ip) const
+	{
+		return (ip.type == type) && (ip.row == row);
+	}
+};
+
+QList<ItemPos> items;
+
+BreakpointViewer::BreakpointViewer(QWidget* parent)
+	: QTabWidget(parent),
+	  ui(new Ui::BreakpointViewer)
+{
+	setupUi(this);
+	bpTableWidget->sortByColumn(BP_ADDRESS, Qt::AscendingOrder);
+	wpTableWidget->sortByColumn(WP_REGION, Qt::AscendingOrder);
+	bpTableWidget->setColumnHidden(WP_TYPE, true);
+	bpTableWidget->resizeColumnsToContents();
+	wpTableWidget->resizeColumnsToContents();
+
+	selector[0] = bpTableWidget;
+	selector[1] = wpTableWidget;
+	userMode = true;
+
+	connect(bpTableWidget, &QTableWidget::itemChanged, this,
+			&BreakpointViewer::changeBpTableItem);
+	connect(wpTableWidget, &QTableWidget::itemChanged, this,
+			&BreakpointViewer::changeWpTableItem);
+}
+
+// TODO: move the createSetCommand to a session manager
+void BreakpointViewer::createBreakpoint(Type type, int row)
+{
+	// qDebug() << "createBreakpoint() called";
+	auto table = selector[type];
+	Breakpoints::Type type2;
+
+	if (type == WATCHPOINT) {
+		auto model = table->model();
+		auto combo = (QComboBox*) table->indexWidget(model->index(row, WP_TYPE));
+		type2 = static_cast<Breakpoints::Type>(combo->currentIndex() + 1);
+	} else {
+		type2 = Breakpoints::BREAKPOINT;
+	}
+
+	// qDebug() << "createBreakpoint, type" << type << "row" << row;
+	// for (auto item: items) {
+	// 	qDebug() << "item" << item.type << item.row;
+	// }
+
+	bool ok;
```

Comment:
No need to declare `ok` upfront. I understand this makes the line `ok = process...` more similar to the lines that follow. But in general try to avoid declaring variables without also initializing them.